### ISSUE TITLE
Last batch of counter metrics are lost when using StepCounter

### ISSUE THREAD

**Issue by mbbeaubi** [elastic/apm-agent-java#3000]

**Comment by jackshirazi** [comment]
I assume you mean a micrometer StepCounter? This is how micrometer works, it only provides the updated value after the step interval

**Comment by mbbeaubi** [comment]
Yes, I am referring to micrometer.

I understand this is the behavior of StepCounter in micrometer, but the reporter (sorry if I am using the wrong word here) pushing the metrics to the APM server is part of the APM java agent.

The apm agent reporter understands that the program is being shutdown, that's why you have the hook to publish metrics one final time. That hook should also take care of pushing the current values of the meter, otherwise metrics will be lost.

**Comment by jackshirazi** [comment]
The agent *is* reporting the current value from Micrometer at shutdown. Micrometer is implemented with eventual consistency so that only at reporting time - which is defined by the user with the interval they set - is the value updated. There is no mechanism to obtain the value (which any other thread may have updated and which is not necessarily visible to the reporting thread) until the interval threshold has been crossed. 

(Not strictly relevant to this discussion is that the micrometer interval threshold is an arbitrary point within the first interval, and then at that same point each susbsequent interval).

The upshot is that the agent reports exactly what you would print if you asked the counter for the value at that time. For us to report a different "more up to date" value would be extremely difficult - and more importantly would be a bug

**Comment by mbbeaubi** [comment]
I get your point about this being mostly a micrometer bug. In fact here is their git issue opened 3 years ago:

https://github.com/micrometer-metrics/micrometer/issues/1882

You are correct that without better support from them, any fix APM could make would be not reliable. However, since I am using APM to ship my metrics, I at least need some sort of hook into the APM shutdown process so that I can do it correctly.

If I was using the micrometer provided ElasticMeterRegistry, I could take the patch given in that issue and hack together a solution that would work. However, since APM is magically shipping all my metrics outside my control, I have no ability to fix this issue even if I was willing to code, test and support the fix myself.

APM is used specifically for shipping metrics to elastic, APM recommends we all use step style metrics because that works best with kibana. Even if APM can't directly support a fix because there is no official way to do it coming from micrometer, I'm sure a lot of users would be willing to do it themselves with some sort of plugin if APM gave us that ability.

**Comment by jackshirazi** [comment]
Well from your initial description above, maybe just try the cumulative option instead of step?

**Comment by mbbeaubi** [comment]
Cumulative doesn't work well with kibana reports

**Comment by mbbeaubi** [comment]
So there's another bug with apm java agents and metrics I realized.

I put a sleep in at the end of my program equal to the step interval to ensure my metrics got pushed. They got pushed twice. This is what I'm thinking is happening, my step interval is 5 seconds.

Second 0: I write a value to a new counter
Second 0.5: APM scheduled thread wakes up, nothing to push (need to wait 5 seconds before value will be available)
Second 1: I sleep for 5 seconds
Second 5.5: APM scheduled thread wakes up and pushes the value to the server
Second 6: Main method sleep wakes up, program shutdown begins
Second 6.1: APM finalizer thread pushes the same value to the server again

Seems to me this would happen all the time to people using step metrics and APM. Is there anyway to cancel the APM meter push during the finalizer thread? Seems to me like it should just be disabled automatically if step mode is enabled for a registry since the behavior is so unpredictable.

The only way I can think to resolve is to sleep for 2x the step interval, which just seems crazy.

**Comment by jackshirazi** [comment]
Yes this latter is a bug, thanks for reporting it. Basically the shutdown should only report cumulative metrics and not interval ones since the interval is now mid-interval (otherwise the reporter would have already reported it). We'll have a fix for that in the next release

**Comment by jackshirazi** [comment]
@mbbeaubi please can you confirm which agent version you are using

**Comment by jackshirazi** [comment]
Thanks, please can you try the latest version (1.36.0) and confirm whether you see the same behaviour

**Comment by jackshirazi** [comment]
Without confirmation this is still an issue, I'll be closing this in the next couple of days (of course it can always be reopened later)
